File: SCALA/Factorial/Factorial.scala
/**
  * Created by @amanmehara on 24-09-2017.
  */
object Factorial {

  def factorial (number: Int, accumulator: BigInt = 1): BigInt = {
    if(number > 0) factorial(number - 1, accumulator * number) else accumulator
  }

}

File: JAVA/Factorial/Factorial.java
import java.math.BigInteger;

public class Factorial {

    public BigInteger factorial(Integer number) {
        return factorial(number, BigInteger.ONE);
    }

    private BigInteger factorial(Integer number, BigInteger accumulator) {
        return number > 0 ? factorial(number - 1, accumulator.multiply(BigInteger.valueOf(number))) : accumulator;
    }

}